cpp/src/arrow/util/future_test.cc

```diff
@@ -122,15 +122,21 @@ void AssertFinished(const Future<T>& fut) {
 // Assert the future is successful *now*
 template <typename T>
 void AssertSuccessful(const Future<T>& fut) {
-  ASSERT_EQ(fut.state(), FutureState::SUCCESS);
-  ASSERT_OK(fut.status());
+  ASSERT_TRUE(fut.Wait(0.1));
```

Comment:
Hmm... this doesn't match the comment above anymore.

Reply:
Perhaps instead add separate `AssertSuccessfulSoon` and `AssertFailedSoon`?

Reply:
Done

Reply:
I should clarify, I took out the `Wait(0.1)` since the check beforehand that it was completed was good enough.  So now if it is in the pending state it will fail.  This change was intended to make debugging easier.  `AssertSuccessful` calls `fut.state()` which (if the future is pending) will wait on the pending future.  This means a bug in the test or code which resulted in a pending future being passed to `AssertSuccessful` would deadlock and this was making the tests difficult to debug.

Now we simply check for pending first as a special case and fail if it is pending before we check the state.

Reply:
Hmm, I think this is a bit mixed up. `Future::state` is not blocking, `Future::status` is. But since we're first asserting that the future state is completed by calling `Future::state` and asserting on the result, `AssertSuccessful` could not succeed if the future is pending.

Reply:
Yes, that is write, `status` was the blocking call.  You are also right that it was not succeeding.  However, it would hang indefinitely.  This made it more difficult to debug.

So the old implementation was correct, but inconvenient.